server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/cics/utility/CICSReadqOptionsCheckUtility.java

```diff
@@ -60,36 +59,41 @@ public CICSReadqOptionsCheckUtility(
      * @param <E> A subclass of ParserRuleContext
      */
     public <E extends ParserRuleContext> void checkOptions(E ctx) {
-        if (ctx.getRuleIndex() == RULE_cics_readq_td)
-            checkTd((CICSParser.Cics_readq_tdContext) ctx);
-        else if (ctx.getRuleIndex() == RULE_cics_readq_ts)
-            checkTs((CICSParser.Cics_readq_tsContext) ctx);
+        if (ctx.getRuleIndex() == RULE_cics_readq_ts_td) {
+            checkOpts((CICSParser.Cics_readq_ts_tdContext) ctx);
+        }
 
         checkDuplicates(ctx);
     }
 
-    private void checkTd(CICSParser.Cics_readq_tdContext ctx) {
-        if (!ctx.TD().isEmpty())
-            checkHasMandatoryOptions(ctx.QUEUE(), ctx, "QUEUE");
-        if (!ctx.NOSUSPEND().isEmpty())
-            checkHasMandatoryOptions(ctx.TD(), ctx, "TD");
-        if (!ctx.cics_into_set().isEmpty())
-            checkSetTd(ctx.cics_into_set().listIterator().next());
-
+    private void checkOpts(CICSParser.Cics_readq_ts_tdContext ctx) {
+       if (checkHasMutuallyExclusiveOptions("TD or TS", ctx.TD(), ctx.TS()) != 2) {
+           if (!ctx.TD().isEmpty()) {
+               checkTd(ctx);
+           } else {
+               checkTs(ctx);
+           }
+       }
```

Comment:
If I am reading the code right the return value is just a number of nodes, not the number of distinct types of nodes. So I suggest just
```suggestion
       final List<TerminalNode> TDs = ctx.TD();
       final List<TerminalNode> TSs = ctx.TS();
       checkHasMutuallyExclusiveOptions("TD or TS", TDs, TSs);
       if (TDs.isEmpty())
           checkTs(ctx);
       else if (TSs.isEmpty())
           checkTd(ctx);
```

Reply:
It is checking if TD and TS are present together. Other checks would be redundant in this condition. Distinct types should not a concern hence duplication checks going to be performed.

Reply:
`TD TD` gives just a warning, but `checkHasMutuallyExclusiveOptions` will return 2 preventing checking of the rest of the arguments.